### PR TITLE
Implement basic DSM user admin service and endpoints

### PR DESCRIPTION
**Pull request by denniscunningham**
PEPPER-970

The DSM user administration feature is described in [DSM user administration for study staff.](https://docs.google.com/document/d/1PNYMTRj8eE1XekNk4ejTIGBp353d3240cAgBJ19MvG8/edit) That document also links to other documents including [DSM user administration requests](https://docs.google.com/document/d/1RK8qeOY48Rt5cslgdhO5LrtN2C88im3uRugzj_3KkcI/edit?usp=sharing), which describes the HTTP request and response formats.

See the PEPPER-963 other sub-tasks to see what this PR is not intended to cover.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserInfo.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserInfo.java]: Since we are not adding to any existing roles and setting the roles passed, maybe we can call it setRoles(List<UserRole> roles)

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRequest.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRequest.java:11]: We can use @AllArgsConstructor and avoid this constructor
ex: 
@Data
@AllArgsConstructor
public class UserRequest {

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRole.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRole.java:11]: We can use @AllArgsConstructor and get rid of this constructor

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRoleRequest.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRoleRequest.java:13]: We can use @AllArgsConstructor and get rid of this constructor

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRoleResponse.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserRoleResponse.java:9]: Might not gain much but alternate way to reduce lines of code.
use  @NoArgsConstructor and initialize users in the declaration

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java]: curious why "" rather than null ?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java:24]: We can use @AllArgsConstructor and get rid of this constructor

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/StudyRoleResponse.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/StudyRoleResponse.java:12]: We can use @AllArgsConstructor and get rid of this constructor . Initialize roles in the declaration

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/StudyRoleResponse.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/StudyRoleResponse.java:29]: We can use @AllArgsConstructor and get rid of this constructor

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java]: Why is this method empty ?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:803]: We can use @AllArgsConstructor and get rid of the constructor

**Review by ssettipalli — COMMENTED** [review]
In most of the DTO's we can use can use @AllArgsConstructor and get rid of the constructor reducing the code.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java]: inTransaction seem to be deprecated, we can still use it though.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserInfo.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserInfo.java]: Good point. Will change.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java]: We use it everywhere in DSM... I have yet to see any code using the alternative. 
I'll do some research and add this to our tech debt discussion. I suspect the deprecation is incorrect.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java]: I am typically reluctant to change existing code given the low DSM code coverage, but I'll look more closely at this. This method is called sparingly, so the fan out from changes is low. Since email is required, that part is at least incorrect.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java:24]: I'm not a fan of AllArgsConstructor since it lowers the readability of the code. Why do you feel it is beneficial?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java]: Cruft. Will remove.

**Review by denniscunningham — COMMENTED** [review]
Thx for your review Sampath!
See my comments.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java]: we do have quiet a bit of deprecated stuff in use anyway

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java:24]: looks clean and less code, thats all.

**Review by ssettipalli — APPROVED** [review]
looks ok to me.
May be @pegahtah can take another deeper look into functionality.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/AddUserRequest.java:24]: Thx. You may have noticed that I used it for the updated UserDto. It works well there.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:225]: Am I correct that this is checking that all the roles do belong to this study group?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:149]: Is this method only getting the newly added roles for the participant? To avoid adding a role multiple times for a user

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:154]: Shouldn't we check for this before the for? In order to avoid half of the roles getting inserted before finding a blank role in the middle of `roles`

Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:154]: Yes! I actually improved that in my follow-up branch (for editing user info). 
Is it okay if I leave that code in the other branch (it will be ready to merge once this is approved), or would you prefer to have it in this branch?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:225]: 'validRoleNames' are all the roles the operator/admin can manage. If the operator is `pepper_admin` then that is all the study roles. If the operator is `study_user_admin` then it is less than all the study roles -- for instance, no GP only roles, no `study_user_admin`, etc.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:149]: This method adds the requested roles (`roles`) for one user (`email`), and validates the roles against `studyRoles`, which also has the role ID (so we don't have to ask the DB for it again). As you noted below, this is improved in the upcoming branch to validate the requested roles before calling `addRoles`.

**Review by denniscunningham — COMMENTED** [review]
@pegahtah Thx for your review! Let me know if my answers make sense, or if you have follow-up questions.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:149]: I see! Thank you for explaining that!
One thing I am still confused about (sorry if it sounds too much overthinking, probably because I have not see the frontend and how it will be handled) is if user `A` already has the` kit_shipping` role, and in frontend the admin also assigns the `kit_deactivation` role for them, are both of those roles coming to `addRoles` in the `roles` List and the `kit_shipping` is then added twice? I think there should be a check for what is a "new" role vs an existing role of a user, either from frontend or backend to stops that, but I think I missed it here.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:154]: Yes Thank you!

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java:149]: Ah, I think I answered the wrong question initially. 
If I understand correctly, you are asking what prevents the code from adding the same role for a user twice, which is a valid concern since there is no key constraint to prevent that (and thus we have users with more than one instance of the same role, at least on Dev).
We do that check at the beginning of addUserRole, which is the only place we add user roles.

Does that answer your question, or did I misunderstand again?

**Review by pegahtah — APPROVED** [review]
LGTM! Thank you for answering my questions

**Comment by denniscunningham** [comment]
@pegahtah Thx for your review. I love the questions... that's what the review is for!

**Pull request merged**
